File: src/main/python/g0201_0300/s0283_move_zeroes/Solution.py
# #Easy #Top_100_Liked_Questions #Array #Two_Pointers #LeetCode_75_Two_Pointers
# #Algorithm_I_Day_3_Two_Pointers #Programming_Skills_I_Day_6_Array #Udemy_Arrays
# #Big_O_Time_O(n)_Space_O(1) #2025_07_25_Time_3_ms_(80.05%)_Space_18.63_MB_(94.75%)

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        firstZero = 0
        for i in range(len(nums)):
            if nums[i] != 0:
                self.swap(firstZero, i, nums)
                firstZero += 1

    def swap(self, index1: int, index2: int, nums: List[int]) -> None:
        nums[index1], nums[index2] = nums[index2], nums[index1]

File: src/main/ruby/g0201_0300/s0283_move_zeroes/solution.rb
# #Easy #Top_100_Liked_Questions #Array #Two_Pointers #LeetCode_75_Two_Pointers
# #Algorithm_I_Day_3_Two_Pointers #Programming_Skills_I_Day_6_Array #Udemy_Arrays
# #Big_O_Time_O(n)_Space_O(1) #2023_11_26_Time_95_ms_(78.01%)_Space_212.9_MB_(32.46%)

# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def move_zeroes(nums)
  first_zero = 0

  (0...nums.length).each do |i|
    if nums[i] != 0
      swap(first_zero, i, nums)
      first_zero += 1
    end
  end
end

private

def swap(index1, index2, numbers)
  val2 = numbers[index2]
  numbers[index2] = numbers[index1]
  numbers[index1] = val2
end
